src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs

```diff
@@ -99,9 +108,24 @@ internal bool FindNewWork(out long nFromInclusiveLocal, out long nToExclusiveLoc
                         Interlocked.CompareExchange(ref _indexRanges[_nCurrentIndexRange]._nSharedCurrentIndexOffset, new Box<long>(0), null);
                     }
 
-                    // this access needs to be on the array slot
-                    long nMyOffset = Interlocked.Add(ref _indexRanges[_nCurrentIndexRange]._nSharedCurrentIndexOffset.Value,
-                                                    _nIncrementValue) - _nIncrementValue;
+                    long nMyOffset;
+                    if (IntPtr.Size == 4 && _use32BitCurrentIndex)
+                    {
+                        // In 32-bit processes, we prefer to use 32-bit interlocked operations, to avoid the possibility of doing
+                        // a 64-bit interlocked when the target value crosses a cache line, as that can be super expensive.
+                        // We use the first 32 bits of the Int64 index in such cases.
+                        unsafe
+                        {
+                            fixed (long* indexPtr = &_indexRanges[_nCurrentIndexRange]._nSharedCurrentIndexOffset.Value)
```

Comment:
Could you use ref locals here instead of pinning (of course you're just replacing unsafe with the Unsafe library to do the cast).

```C#
ref int index = ref Unsafe.As<long, int>(ref currentRange._nSharedCurrentIndexOffset.Value);
nMyOffset = Interlocked.Add(ref index, (int)_nIncrementValue) - _nIncrementValue;
```